src/renderer/caching_shaper.rs

```diff
@@ -48,101 +40,159 @@ struct Asset;
 
 const DEFAULT_FONT_SIZE: f32 = 14.0;
 
+#[derive(Clone)]
+pub struct ExtendedFontFamily {
+    pub fonts: Vec<SkriboFont>,
+}
+
+impl ExtendedFontFamily {
+    pub fn new() -> ExtendedFontFamily {
+        ExtendedFontFamily { fonts: Vec::new() }
+    }
+
+    pub fn add_font(&mut self, font: SkriboFont) {
+        self.fonts.push(font);
+    }
+
+    pub fn get(&self, props: Properties) -> Option<&Font> {
+        for handle in &self.fonts {
+            let font = &handle.font;
+            let properties = font.properties();
+
+            if properties.weight == props.weight && properties.style == props.style {
+                return Some(&font);
+            }
+        }
+
+        if let Some(handle) = &self.fonts.first() {
+            return Some(&handle.font);
+        }
+
+        None
+    }
+
+    pub fn to_family(self) -> FontFamily {
+        let mut new_family = FontFamily::new();
+
+        for font in self.fonts {
+            new_family.add_font(font);
+        }
+
+        new_family
+    }
+}
+
+pub struct FontLoader {
+    cache: LruCache<String, ExtendedFontFamily>,
+    source: SystemSource,
+}
+
+impl FontLoader {
+    pub fn new() -> FontLoader {
+        FontLoader {
+            cache: LruCache::new(10),
+            source: SystemSource::new(),
+        }
+    }
+
+    fn get(&mut self, font_name: &str) -> Option<ExtendedFontFamily> {
+        if let Some(family) = self.cache.get(&String::from(font_name)) {
+            return Some(family.clone());
+        }
+
+        None
+    }
+
+    #[cfg(feature = "embed-fonts")]
+    fn load_from_asset(&mut self, font_name: &str) -> Option<ExtendedFontFamily> {
+        let mut family = ExtendedFontFamily::new();
+
+        Asset::get(font_name)
+            .and_then(|font_data| Font::from_bytes(font_data.to_vec().into(), 0).ok())
+            .map(|font| family.add_font(SkriboFont::new(font)));
+
+        self.cache.put(String::from(font_name), family);
+        self.get(font_name)
+    }
+
+    fn load(&mut self, font_name: &str) -> Option<ExtendedFontFamily> {
+        let handle = match self.source.select_family_by_name(font_name) {
+            Ok(it) => it,
+            _ => return None,
+        };
+
+        let fonts = handle.fonts();
+        let mut family = ExtendedFontFamily::new();
```

Comment:
minor: maybe it would make sense to add a function `from_normal_font_family` as a "static" function to ExtendedFontFamily to simplify the logic here? Maybe not worth the effort.

Reply:
nah, it was worth, more clean